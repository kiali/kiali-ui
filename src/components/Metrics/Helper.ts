import assign from 'lodash/fp/assign';
import {
  MonitoringDashboard,
  AllLabelsValues,
  SingleLabelValues,
  LabelDisplayName,
  PromLabel,
  Aggregation,
  AllPromLabelsValues,
  TimeSeries
} from '../../types/Metrics';
import { BaseMetricsOptions } from '../../types/MetricsOptions';
import { MetricsSettingsDropdown, MetricsSettings } from '../MetricsOptions/MetricsSettings';
import MetricsDuration from '../MetricsOptions/MetricsDuration';
import { DurationInSeconds } from '../../types/Common';
import { computePrometheusRateParams } from '../../services/Prometheus';

namespace MetricsHelper {
  export const extractLabelValuesOnSeries = (
    series: TimeSeries[],
    aggregations: Aggregation[],
    extracted: AllLabelsValues
  ): void => {
    series.forEach(ts => {
      Object.keys(ts.metric).forEach(k => {
        const agg = aggregations.find(a => a.label === k);
        if (agg) {
          const value = ts.metric[k];
          let values = extracted.get(agg.displayName);
          if (!values) {
            values = {};
            extracted.set(agg.displayName, values);
          }
          values[value] = true;
        }
      });
    });
  };

  export const extractLabelValues = (
    dashboard: MonitoringDashboard,
    previousValues: AllLabelsValues
  ): AllLabelsValues => {
    // Find all labels on all series
    const labelsWithValues: AllLabelsValues = new Map();
    dashboard.aggregations.forEach(agg => labelsWithValues.set(agg.displayName, {}));
    dashboard.charts.forEach(chart => {
      if (chart.metric) {
        extractLabelValuesOnSeries(chart.metric.matrix, dashboard.aggregations, labelsWithValues);
      }
      if (chart.histogram) {
        Object.keys(chart.histogram).forEach(stat => {
          extractLabelValuesOnSeries(chart.histogram![stat].matrix, dashboard.aggregations, labelsWithValues);
        });
      }
    });
    // Keep existing show flag
    labelsWithValues.forEach((values: SingleLabelValues, key: LabelDisplayName) => {
      const previous = previousValues.get(key);
      if (previous) {
        Object.keys(values).forEach(k => {
          if (previous.hasOwnProperty(k)) {
            values[k] = previous[k];
          }
        });
      }
    });
    return labelsWithValues;
  };

  export const mergeLabelFilter = (
    labelValues: AllLabelsValues,
    label: LabelDisplayName,
    value: string,
    checked: boolean
  ): AllLabelsValues => {
    const newLabels = new Map();
    labelValues.forEach((val, key) => {
      const newVal = assign(val)({});
      if (key === label) {
        newVal[value] = checked;
      }
      newLabels.set(key, newVal);
    });
    return newLabels;
  };

  export const convertAsPromLabels = (aggregations: Aggregation[], labels: AllLabelsValues): AllPromLabelsValues => {
    const promLabels = new Map<PromLabel, SingleLabelValues>();
    labels.forEach((val, k) => {
      const chartLabel = aggregations.find(l => l.displayName === k);
      if (chartLabel) {
        promLabels.set(chartLabel.label, val);
      }
    });
    return promLabels;
  };

  export const settingsToOptions = (
    settings: MetricsSettings,
    opts: BaseMetricsOptions,
    aggregations?: Aggregation[]
  ) => {
    opts.avg = settings.showAverage;
    opts.quantiles = settings.showQuantiles;
    opts.byLabels = [];
    if (aggregations) {
      settings.activeLabels.forEach(lbl => {
        const agg = aggregations.find(a => a.displayName === lbl);
        if (agg) {
          opts.byLabels!.push(agg.label);
        }
      });
    }
  };

  export const initMetricsSettings = (opts: BaseMetricsOptions, aggregations?: Aggregation[]) => {
    settingsToOptions(MetricsSettingsDropdown.initialMetricsSettings(), opts, aggregations);
  };

  export const durationToOptions = (duration: DurationInSeconds, opts: BaseMetricsOptions) => {
    opts.duration = duration;
    const intervalOpts = computePrometheusRateParams(duration);
    opts.step = intervalOpts.step;
    opts.rateInterval = intervalOpts.rateInterval;
  };

  export const initDuration = (opts: BaseMetricsOptions): BaseMetricsOptions => {
    durationToOptions(MetricsDuration.initialDuration(), opts);
    return opts;
  };
}

export default MetricsHelper;
