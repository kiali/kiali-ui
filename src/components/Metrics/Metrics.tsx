import * as React from 'react';
import { Alert, Icon } from 'patternfly-react';
import GrafanaInfo from '../../types/GrafanaInfo';
import { authentication } from '../../utils/Authentication';
import * as API from '../../services/Api';
import { computePrometheusQueryInterval } from '../../services/Prometheus';
import MetricsOptionsBar from '../../components/MetricsOptions/MetricsOptionsBar';
import MetricsOptions from '../../types/MetricsOptions';
import HistogramChart from './HistogramChart';
import MetricChart from './MetricChart';
import { Histogram, MetricGroup } from '../../types/Metrics';
import history from '../../app/History';
import { HistoryManager } from '../../app/History';
import { Link } from 'react-router-dom';
import { style } from 'typestyle';

const expandedChartContainerStyle = style({
  height: 'calc(100vh - 248px)'
});

const expandedChartBackLinkStyle = style({
  marginTop: '-1.7em',
  textAlign: 'right'
});

type ChartDefinition = {
  familyName: string;
  component: any;
  metrics?: MetricGroup | Histogram;
};

type MetricsState = {
  alertDetails?: string;
  grafanaLink?: string;
  pollMetrics?: number;
  metricReporter: string;
  metricData?: any;
};

type ObjectId = {
  namespace: string;
  object: string;
};

type MetricsProps = ObjectId & {
  isPageVisible?: boolean;
  objectType: string;
  metricsType: string;
};

class Metrics extends React.Component<MetricsProps, MetricsState> {
  static defaultProps = {
    isPageVisible: true
  };

  options: MetricsOptions;

  constructor(props: MetricsProps) {
    super(props);

    let metricReporter = 'source';

    const metricReporterParam = HistoryManager.getParam('reporter');
    if (metricReporterParam != null) {
      metricReporter = metricReporterParam;
    } else if (this.props.metricsType === 'inbound') {
      metricReporter = 'destination';
    }

    this.state = {
      metricReporter: metricReporter
    };
  }

  getChartsDef(): { [key: string]: ChartDefinition } {
    let inboundCharts = {
      request_count_in: { familyName: 'Request volume (ops)', component: MetricChart },
      request_duration_in: { familyName: 'Request duration (seconds)', component: HistogramChart },
      request_size_in: { familyName: 'Request size (bytes)', component: HistogramChart },
      response_size_in: { familyName: 'Response size (bytes)', component: HistogramChart },
      tcp_received_in: { familyName: 'TCP received (bps)', component: MetricChart },
      tcp_sent_in: { familyName: 'TCP sent (bps)', component: MetricChart }
    };
    let charts: { [key: string]: ChartDefinition } = inboundCharts;

    if (this.props.metricsType === 'outbound') {
      charts = {
        request_count_out: { familyName: 'Request volume (ops)', component: MetricChart },
        request_duration_out: { familyName: 'Request duration (seconds)', component: HistogramChart },
        request_size_out: { familyName: 'Request size (bytes)', component: HistogramChart },
        response_size_out: { familyName: 'Response size (bytes)', component: HistogramChart },
        tcp_received_out: { familyName: 'TCP received (bps)', component: MetricChart },
        tcp_sent_out: { familyName: 'TCP sent (bps)', component: MetricChart }
      };
    }

    if (this.state.metricData) {
      Object.keys(charts).forEach(k => {
        const chart = charts[k];
        const reporter = this.state.metricReporter === 'destination' ? 'dest' : 'source';
        const histo = this.state.metricData[reporter].histograms[k];

        if (histo) {
          chart.metrics = histo;
        } else {
          chart.metrics = this.state.metricData[reporter].metrics[k];
        }
      });
    }

    return charts;
  }

  componentDidMount() {
    this.getGrafanaInfo();
  }

  componentWillUnmount() {
    if (this.state.pollMetrics) {
      clearInterval(this.state.pollMetrics);
    }
  }

  onOptionsChanged = (options: MetricsOptions) => {
    this.options = options;
    options.filters = [
      'request_count',
      'request_size',
      'request_duration',
      'response_size',
      'tcp_received',
      'tcp_sent'
    ];
    const intervalOpts = computePrometheusQueryInterval(options.duration!);
    options.step = intervalOpts.step;
    options.rateInterval = intervalOpts.rateInterval;
    this.fetchMetrics();
    HistoryManager.setParam('duration', options.duration!.toString(10));
  };

  onPollIntervalChanged = (pollInterval: number) => {
    let newRefInterval: number | undefined = undefined;
    if (this.state.pollMetrics) {
      clearInterval(this.state.pollMetrics);
    }
    if (pollInterval > 0) {
      newRefInterval = window.setInterval(this.fetchMetrics, pollInterval);
    }
    this.setState({ pollMetrics: newRefInterval });
    HistoryManager.setParam('pi', pollInterval.toString(10));
  };

  onReporterChanged = (reporter: string) => {
    this.setState({ metricReporter: reporter });
    HistoryManager.setParam('reporter', reporter);
  };

  fetchMetrics = () => {
    let promise;
    switch (this.props.objectType) {
      case 'service':
        promise = API.getServiceMetrics(authentication(), this.props.namespace, this.props.object, this.options);
        break;
      case 'workload':
        promise = API.getWorkloadMetrics(authentication(), this.props.namespace, this.props.object, this.options);
        break;
      case 'app':
        promise = API.getAppMetrics(authentication(), this.props.namespace, this.props.object, this.options);
        break;
      default:
        promise = API.getServiceMetrics(authentication(), this.props.namespace, this.props.object, this.options);
        break;
    }
    Promise.all([promise])
      .then(([response]) => {
        this.setState({ metricData: response.data });
      })
      .catch(error => {
        this.setState({ alertDetails: API.getErrorMsg('Cannot fetch metrics.', error) });
        console.error(error);
      });
  };

  getGrafanaLink(info: GrafanaInfo): string {
    let grafanaLink;
    switch (this.props.objectType) {
      case 'service':
        grafanaLink = `${info.url}${info.serviceDashboardPath}?${info.varService}=${this.props.object}.${
          this.props.namespace
        }.svc.cluster.local`;
        break;
      case 'workload':
        grafanaLink = `${info.url}${info.workloadDashboardPath}?${info.varNamespace}=${this.props.namespace}&${
          info.varWorkload
        }=${this.props.object}`;
        break;
      default:
        grafanaLink = `${info.url}${info.workloadDashboardPath}?${info.varNamespace}=${this.props.namespace}`;
    }
    return grafanaLink;
  }

  getGrafanaInfo = () => {
    API.getGrafanaInfo(authentication())
      .then(response => {
        const info: GrafanaInfo = response['data'];
        if (info) {
          this.setState({
            grafanaLink: this.getGrafanaLink(info)
          });
        } else {
          this.setState({ grafanaLink: undefined });
        }
      })
      .catch(error => {
        this.setState({ grafanaLink: undefined });
        console.error(error);
      });
  };

  dismissAlert = () => this.setState({ alertDetails: undefined });

  render() {
    const urlParams = new URLSearchParams(history.location.search);
    const expandedChart = urlParams.get('expand');
    urlParams.delete('expand');
    const notExpandedLink = history.location.pathname + '?' + urlParams.toString();

    return (
      <div>
        {expandedChart && (
          <h3 className={expandedChartBackLinkStyle}>
            <Link to={notExpandedLink}>
              <Icon name="angle-double-left" /> View all metrics
            </Link>
          </h3>
        )}
        {this.state.alertDetails && <Alert onDismiss={this.dismissAlert}>{this.state.alertDetails}</Alert>}
        <MetricsOptionsBar
          onOptionsChanged={this.onOptionsChanged}
          onPollIntervalChanged={this.onPollIntervalChanged}
          onReporterChanged={this.onReporterChanged}
          onManualRefresh={this.fetchMetrics}
          metricReporter={this.state.metricReporter}
        />
        {expandedChart ? this.renderExpandedChart(expandedChart) : this.renderMetrics()}
      </div>
    );
  }

  renderMetrics() {
    if (!this.props.isPageVisible) {
      return null;
    }

    let charts = this.getChartsDef();

    return (
      <div className="card-pf">
        <div className="row row-cards-pf">
          <div className="col-xs-12">
            <div className="card-pf-accented card-pf-aggregate-status">
              <div className="card-pf-body">
                {Object.keys(charts).map(key => this.renderNormalChart(key, charts[key]))}
              </div>
            </div>
          </div>
          {this.state.grafanaLink && (
            <span id="grafana-link">
              <a href={this.state.grafanaLink} target="_blank">
                View in Grafana
              </a>
            </span>
          )}
        </div>
      </div>
    );
  }

  private renderNormalChart(chartKey: string, chart: ChartDefinition) {
    return this.renderChart(chartKey, chart);
  }

  private renderExpandedChart(chartKey: string) {
    return (
      <div className={expandedChartContainerStyle}>
        {this.renderChart(chartKey, this.getChartsDef()[chartKey], true)}
      </div>
    );
  }

  private renderChart(chartKey: string, chart: ChartDefinition, isExpanded: boolean = false) {
    if (!chart || !chart.metrics) {
      return undefined;
    }
    const props: any = {
      key: chartKey + this.state.metricReporter,
      familyName: chart.familyName
    };
    if ((chart.metrics as MetricGroup).matrix) {
      props.series = (chart.metrics as MetricGroup).matrix;
    } else {
      props.histogram = chart.metrics;
    }

    if (!isExpanded) {
      props.onExpandRequested = () => this.onExpandHandler(chartKey);
    }

    return React.createElement(chart.component, props);
  }

  private onExpandHandler = (chartKey: string): void => {
    const urlParams = new URLSearchParams(history.location.search);
    urlParams.set('expand', chartKey);
    history.push(history.location.pathname + '?' + urlParams.toString());
  };
}

export { MetricsProps };
export default Metrics;
