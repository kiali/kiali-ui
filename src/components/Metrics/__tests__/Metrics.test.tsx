import * as React from 'react';
import { mount, shallow, ReactWrapper } from 'enzyme';

import Metrics from '../Metrics';
import * as API from '../../../services/Api';
import { MetricsDirection, MetricsObjectTypes } from '../../../types/Metrics';

window['SVGPathElement'] = a => a;
let mounted: ReactWrapper<any, any> | null;

const mockAPIToPromise = (func: keyof typeof API, obj: any): Promise<void> => {
  return new Promise((resolve, reject) => {
    jest.spyOn(API, func).mockImplementation(() => {
      return new Promise(r => {
        r({ data: obj });
        setTimeout(() => {
          try {
            resolve();
          } catch (e) {
            reject(e);
          }
        }, 1);
      });
    });
  });
};

const mockServiceMetrics = (metrics: any): Promise<void> => {
  return mockAPIToPromise('getServiceMetrics', { source: metrics, dest: metrics });
};

const mockWorkloadMetrics = (metrics: any): Promise<void> => {
  return mockAPIToPromise('getWorkloadMetrics', { source: metrics, dest: metrics });
};

const mockGrafanaInfo = (info: any): Promise<any> => {
  return mockAPIToPromise('getGrafanaInfo', info);
};

const createMetric = (name: string) => {
  return {
    matrix: [
      {
        metric: { __name__: name },
        values: [[1111, 5], [2222, 10]]
      }
    ]
  };
};

const createHistogram = (name: string) => {
  return {
    average: {
      matrix: [
        {
          metric: { __name__: name },
          values: [[1111, 10], [2222, 11]]
        }
      ]
    },
    median: {
      matrix: [
        {
          metric: { __name__: name },
          values: [[1111, 20], [2222, 21]]
        }
      ]
    },
    percentile95: {
      matrix: [
        {
          metric: { __name__: name },
          values: [[1111, 30], [2222, 31]]
        }
      ]
    },
    percentile99: {
      matrix: [
        {
          metric: { __name__: name },
          values: [[1111, 40], [2222, 41]]
        }
      ]
    }
  };
};

describe('Metrics for a service', () => {
  beforeEach(() => {
    mounted = null;
  });
  afterEach(() => {
    if (mounted) {
      mounted.unmount();
    }
  });

  it('renders initial layout', () => {
    mockGrafanaInfo({});
    const wrapper = shallow(
      <Metrics
        namespace="ns"
        object="svc"
        objectType={MetricsObjectTypes.SERVICE}
        direction={MetricsDirection.INBOUND}
        grafanaInfo={{
          url: 'http://172.30.139.113:3000',
          serviceDashboardPath: '/dashboard/db/istio-dashboard',
          workloadDashboardPath: '/dashboard/db/istio-dashboard',
          varService: 'var-service',
          varNamespace: 'var-namespace',
          varWorkload: 'var-workload'
        }}
      />
    );
    expect(wrapper).toMatchSnapshot();
  });

  it('mounts and loads empty metrics', done => {
    const allMocksDone = [
      mockServiceMetrics({ metrics: {}, histograms: {} })
        .then(() => {
          mounted!.update();
          expect(mounted!.find('.card-pf-body')).toHaveLength(1);
          mounted!.find('.card-pf-body').forEach(pfCard => expect(pfCard.children().length === 0));
        })
        .catch(err => done.fail(err))
    ];
    Promise.all(allMocksDone).then(() => done());
    mounted = mount(
      <Metrics
        namespace="ns"
        object="svc"
        objectType={MetricsObjectTypes.SERVICE}
        direction={MetricsDirection.INBOUND}
        grafanaInfo={{
          url: 'http://172.30.139.113:3000',
          serviceDashboardPath: '/dashboard/db/istio-dashboard',
          workloadDashboardPath: '/dashboard/db/istio-dashboard',
          varService: 'var-service',
          varNamespace: 'var-namespace',
          varWorkload: 'var-workload'
        }}
      />
    );
  });

  it(
    'mounts and loads full metrics',
    done => {
      const allMocksDone = [
        mockServiceMetrics({
          metrics: {
            request_count_in: createMetric('m1')
          },
          histograms: {
            request_size_in: createHistogram('m3'),
            request_duration_in: createHistogram('m5'),
            response_size_in: createHistogram('m7')
          }
        })
          .then(() => {
            mounted!.update();
            expect(mounted!.find('LineChart')).toHaveLength(4);
          })
          .catch(err => done.fail(err))
      ];
      Promise.all(allMocksDone).then(() => done());
      mounted = mount(
        <Metrics
          namespace="ns"
          object="svc"
          objectType={MetricsObjectTypes.SERVICE}
          direction={MetricsDirection.INBOUND}
          grafanaInfo={{
            url: 'http://172.30.139.113:3000',
            serviceDashboardPath: '/dashboard/db/istio-dashboard',
            workloadDashboardPath: '/dashboard/db/istio-dashboard',
            varService: 'var-service',
            varNamespace: 'var-namespace',
            varWorkload: 'var-workload'
          }}
        />
      );
    },
    10000
  ); // Increase timeout for this test
});

describe('Inbound Metrics for a workload', () => {
  beforeEach(() => {
    mounted = null;
  });
  afterEach(() => {
    if (mounted) {
      mounted.unmount();
    }
  });

  it('renders initial layout', () => {
    const wrapper = shallow(
      <Metrics
        namespace="ns"
        object="svc"
        objectType={MetricsObjectTypes.WORKLOAD}
        direction={MetricsDirection.INBOUND}
        grafanaInfo={{
          url: 'http://172.30.139.113:3000',
          serviceDashboardPath: '/dashboard/db/istio-dashboard',
          workloadDashboardPath: '/dashboard/db/istio-dashboard',
          varService: 'var-service',
          varNamespace: 'var-namespace',
          varWorkload: 'var-workload'
        }}
      />
    );
    expect(wrapper).toMatchSnapshot();
  });

  it('mounts and loads empty metrics', done => {
    const allMocksDone = [
      mockWorkloadMetrics({ metrics: {}, histograms: {} })
        .then(() => {
          mounted!.update();
          expect(mounted!.find('.card-pf-body')).toHaveLength(1);
          mounted!.find('.card-pf-body').forEach(pfCard => expect(pfCard.children().length === 0));
        })
        .catch(err => done.fail(err))
    ];
    Promise.all(allMocksDone).then(() => done());
    mounted = mount(
      <Metrics
        namespace="ns"
        object="svc"
        objectType={MetricsObjectTypes.WORKLOAD}
        direction={MetricsDirection.INBOUND}
        grafanaInfo={{
          url: 'http://172.30.139.113:3000',
          serviceDashboardPath: '/dashboard/db/istio-dashboard',
          workloadDashboardPath: '/dashboard/db/istio-dashboard',
          varService: 'var-service',
          varNamespace: 'var-namespace',
          varWorkload: 'var-workload'
        }}
      />
    );
  });

  it(
    'mounts and loads full metrics',
    done => {
      const allMocksDone = [
        mockWorkloadMetrics({
          metrics: {
            request_count_in: createMetric('m1')
          },
          histograms: {
            request_size_in: createHistogram('m3'),
            request_duration_in: createHistogram('m5'),
            response_size_in: createHistogram('m7')
          }
        })
          .then(() => {
            mounted!.update();
            expect(mounted!.find('LineChart')).toHaveLength(4);
          })
          .catch(err => done.fail(err))
      ];
      Promise.all(allMocksDone).then(() => done());
      mounted = mount(
        <Metrics
          namespace="ns"
          object="svc"
          objectType={MetricsObjectTypes.WORKLOAD}
          direction={MetricsDirection.INBOUND}
          grafanaInfo={{
            url: 'http://172.30.139.113:3000',
            serviceDashboardPath: '/dashboard/db/istio-dashboard',
            workloadDashboardPath: '/dashboard/db/istio-dashboard',
            varService: 'var-service',
            varNamespace: 'var-namespace',
            varWorkload: 'var-workload'
          }}
        />
      );
    },
    10000
  ); // Increase timeout for this test
});

describe('Outbound Metrics for a workload', () => {
  beforeEach(() => {
    mounted = null;
  });
  afterEach(() => {
    if (mounted) {
      mounted.unmount();
    }
  });

  it('renders initial layout', () => {
    const wrapper = shallow(
      <Metrics
        namespace="ns"
        object="svc"
        objectType={MetricsObjectTypes.WORKLOAD}
        direction={MetricsDirection.INBOUND}
        grafanaInfo={{
          url: 'http://172.30.139.113:3000',
          serviceDashboardPath: '/dashboard/db/istio-dashboard',
          workloadDashboardPath: '/dashboard/db/istio-dashboard',
          varService: 'var-service',
          varNamespace: 'var-namespace',
          varWorkload: 'var-workload'
        }}
      />
    );
    expect(wrapper).toMatchSnapshot();
  });

  it('mounts and loads empty metrics', done => {
    const allMocksDone = [
      mockWorkloadMetrics({ metrics: {}, histograms: {} })
        .then(() => {
          mounted!.update();
          expect(mounted!.find('.card-pf-body')).toHaveLength(1);
          mounted!.find('.card-pf-body').forEach(pfCard => expect(pfCard.children().length === 0));
        })
        .catch(err => done.fail(err))
    ];
    Promise.all(allMocksDone).then(() => done());
    mounted = mount(
      <Metrics
        namespace="ns"
        object="svc"
        objectType={MetricsObjectTypes.WORKLOAD}
        direction={MetricsDirection.INBOUND}
        grafanaInfo={{
          url: 'http://172.30.139.113:3000',
          serviceDashboardPath: '/dashboard/db/istio-dashboard',
          workloadDashboardPath: '/dashboard/db/istio-dashboard',
          varService: 'var-service',
          varNamespace: 'var-namespace',
          varWorkload: 'var-workload'
        }}
      />
    );
  });

  it(
    'mounts and loads full metrics',
    done => {
      const allMocksDone = [
        mockWorkloadMetrics({
          metrics: {
            request_count_out: createMetric('m1')
          },
          histograms: {
            request_size_out: createHistogram('m3'),
            request_duration_out: createHistogram('m5'),
            response_size_out: createHistogram('m7')
          }
        })
          .then(() => {
            mounted!.update();
            expect(mounted!.find('LineChart')).toHaveLength(4);
          })
          .catch(err => done.fail(err))
      ];
      Promise.all(allMocksDone).then(() => done());
      mounted = mount(
        <Metrics
          namespace="ns"
          object="svc"
          objectType={MetricsObjectTypes.WORKLOAD}
          direction={MetricsDirection.OUTBOUND}
          grafanaInfo={{
            url: 'http://172.30.139.113:3000',
            serviceDashboardPath: '/dashboard/db/istio-dashboard',
            workloadDashboardPath: '/dashboard/db/istio-dashboard',
            varService: 'var-service',
            varNamespace: 'var-namespace',
            varWorkload: 'var-workload'
          }}
        />
      );
    },
    10000
  ); // Increase timeout for this test
});
