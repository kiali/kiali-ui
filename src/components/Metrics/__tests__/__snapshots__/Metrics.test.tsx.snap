// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Inbound Metrics for a workload renders initial layout 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Provider
    store={
      Object {
        "dispatch": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
        Symbol(observable): [Function],
      }
    }
  >
    <Metrics
      direction={0}
      grafanaInfo={
        Object {
          "serviceDashboardPath": "/dashboard/db/istio-dashboard",
          "url": "http://172.30.139.113:3000",
          "varNamespace": "var-namespace",
          "varService": "var-service",
          "varWorkload": "var-workload",
          "workloadDashboardPath": "/dashboard/db/istio-dashboard",
        }
      }
      isPageVisible={true}
      namespace="ns"
      object="svc"
      objectType={1}
    />
  </Provider>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "direction": 0,
      "grafanaInfo": Object {
        "serviceDashboardPath": "/dashboard/db/istio-dashboard",
        "url": "http://172.30.139.113:3000",
        "varNamespace": "var-namespace",
        "varService": "var-service",
        "varWorkload": "var-workload",
        "workloadDashboardPath": "/dashboard/db/istio-dashboard",
      },
      "isPageVisible": true,
      "namespace": "ns",
      "object": "svc",
      "objectType": 1,
    },
    "ref": null,
    "rendered": null,
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "direction": 0,
        "grafanaInfo": Object {
          "serviceDashboardPath": "/dashboard/db/istio-dashboard",
          "url": "http://172.30.139.113:3000",
          "varNamespace": "var-namespace",
          "varService": "var-service",
          "varWorkload": "var-workload",
          "workloadDashboardPath": "/dashboard/db/istio-dashboard",
        },
        "isPageVisible": true,
        "namespace": "ns",
        "object": "svc",
        "objectType": 1,
      },
      "ref": null,
      "rendered": null,
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;

exports[`Metrics for a service renders initial layout 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Provider
    store={
      Object {
        "dispatch": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
        Symbol(observable): [Function],
      }
    }
  >
    <Metrics
      direction={0}
      grafanaInfo={
        Object {
          "serviceDashboardPath": "/dashboard/db/istio-dashboard",
          "url": "http://172.30.139.113:3000",
          "varNamespace": "var-namespace",
          "varService": "var-service",
          "varWorkload": "var-workload",
          "workloadDashboardPath": "/dashboard/db/istio-dashboard",
        }
      }
      isPageVisible={true}
      namespace="ns"
      object="svc"
      objectType={0}
    />
  </Provider>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "direction": 0,
      "grafanaInfo": Object {
        "serviceDashboardPath": "/dashboard/db/istio-dashboard",
        "url": "http://172.30.139.113:3000",
        "varNamespace": "var-namespace",
        "varService": "var-service",
        "varWorkload": "var-workload",
        "workloadDashboardPath": "/dashboard/db/istio-dashboard",
      },
      "isPageVisible": true,
      "namespace": "ns",
      "object": "svc",
      "objectType": 0,
    },
    "ref": null,
    "rendered": null,
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "direction": 0,
        "grafanaInfo": Object {
          "serviceDashboardPath": "/dashboard/db/istio-dashboard",
          "url": "http://172.30.139.113:3000",
          "varNamespace": "var-namespace",
          "varService": "var-service",
          "varWorkload": "var-workload",
          "workloadDashboardPath": "/dashboard/db/istio-dashboard",
        },
        "isPageVisible": true,
        "namespace": "ns",
        "object": "svc",
        "objectType": 0,
      },
      "ref": null,
      "rendered": null,
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;

exports[`Outbound Metrics for a workload renders initial layout 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Provider
    store={
      Object {
        "dispatch": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
        Symbol(observable): [Function],
      }
    }
  >
    <Metrics
      direction={0}
      grafanaInfo={
        Object {
          "serviceDashboardPath": "/dashboard/db/istio-dashboard",
          "url": "http://172.30.139.113:3000",
          "varNamespace": "var-namespace",
          "varService": "var-service",
          "varWorkload": "var-workload",
          "workloadDashboardPath": "/dashboard/db/istio-dashboard",
        }
      }
      isPageVisible={true}
      namespace="ns"
      object="svc"
      objectType={1}
    />
  </Provider>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "direction": 0,
      "grafanaInfo": Object {
        "serviceDashboardPath": "/dashboard/db/istio-dashboard",
        "url": "http://172.30.139.113:3000",
        "varNamespace": "var-namespace",
        "varService": "var-service",
        "varWorkload": "var-workload",
        "workloadDashboardPath": "/dashboard/db/istio-dashboard",
      },
      "isPageVisible": true,
      "namespace": "ns",
      "object": "svc",
      "objectType": 1,
    },
    "ref": null,
    "rendered": null,
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "direction": 0,
        "grafanaInfo": Object {
          "serviceDashboardPath": "/dashboard/db/istio-dashboard",
          "url": "http://172.30.139.113:3000",
          "varNamespace": "var-namespace",
          "varService": "var-service",
          "varWorkload": "var-workload",
          "workloadDashboardPath": "/dashboard/db/istio-dashboard",
        },
        "isPageVisible": true,
        "namespace": "ns",
        "object": "svc",
        "objectType": 1,
      },
      "ref": null,
      "rendered": null,
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
